### PR TITLE
alter some interfaces to support GameEngine testability

### PR DESCRIPTION
**Pull request by kaen**
### Contains

Interface changes to support ModuleTestingEnvironment (creation and manipulation of TerasologyEngines)
see https://github.com/kaen/ModuleTestingEnvironment

### How to test

Engine should work exactly as it did before :)

### CODE REVIEW

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by Cervator** [comment]
Neat! Pinging @immortius @flo @skaldarnar to maybe get some quick review :-)

File: engine/src/main/java/org/terasology/engine/TerasologyEngine.java [hunk 1]
```diff
@@ -389,45 +389,56 @@ public synchronized void run(GameState initialState) {
...
+        PerformanceMonitor.endActivity();
+    }
 
-            processPendingState();
+    /**
+     * Runs a single "tick" of the engine
+     * @return true if the loop requesting a tick should continue running
+     */
+    public boolean tick() {
+        if(shutdownRequested) {
```
Comment (immortius) [inline engine/src/main/java/org/terasology/engine/TerasologyEngine.java]: Pedantic comment, but missing a space between if and bracket.

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by bkconrad** [comment]
Getting a client to connect to a host required some additional changes. Chiefly I had to normalize the behavior of States and Engines regarding when CoreRegistry is set. Now by definition a State is required to expose a context, and that context will be set as CoreRegistry when the engine is `tick`d.

Additionally, NetworkSystem now uses a local context reference rather than CoreRegistry in order to make it thread-safe (since some parts are apparently executed off of the main thread). The engine now sets the NetworkSystem's local contex on each `tick`. This is equivalent to the old behavior of relying on the State to set its own context as CoreRegistry, but without the CoreRegistry reference it becomes thread safe in the test environment. engine-tests have been updated to reflect this requirement.

Additionally I exposed the cleanup method to allow proper shutdown/cleanup of the Engine instances between test cases

**Comment by kaen** [comment]
Getting a client to connect to a host required some additional changes. Chiefly I had to normalize the behavior of States and Engines regarding when CoreRegistry is set. Now by definition a State is required to expose a context, and that context will be set as CoreRegistry when the engine is `tick`d.

Additionally, NetworkSystem now uses a local context reference rather than CoreRegistry in order to make it thread-safe (since some parts are apparently executed off of the main thread). The engine now sets the NetworkSystem's local contex on each `tick`. This is equivalent to the old behavior of relying on the State to set its own context as CoreRegistry, but without the CoreRegistry reference it becomes thread safe in the test environment. engine-tests have been updated to reflect this requirement.

Additionally I exposed the cleanup method to allow proper shutdown/cleanup of the Engine instances between test cases

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by Cervator** [comment]
Tested it out, both in single, server-client + client, and headless + client, seemed solid! Well, found a pre-existing bug I hadn't noticed before, but that's something else :-)

Thanks!

**Pull request merged**